File: android/app/src/main/java/dev/pepotech/pepomote/control/PressStep.kt
package dev.pepotech.pepomote.control

/**
 * Qué hace un dedo apoyado cuando se mueve por encima de los botones, según
 * los dos ajustes de pulsación. Función pura, la misma en las tres apps (como
 * [DpadModel] y `dpad.rs`): se prueba en la JVM.
 *
 * - «Pulsar deslizando» ([slide]): manda lo que hay bajo el dedo; salirse de
 *   un botón lo suelta y entrar en otro lo pulsa, aunque el dedo naciera en
 *   el vacío.
 * - «Mantener al salir del botón» ([sticky], de serie): lo que se pulsó
 *   sigue pulsado hasta levantar el dedo, aunque se salga del botón.
 * - Sin ninguno de los dos: salirse del botón lo suelta y el dedo se queda
 *   libre; ya no vuelve a coger nada hasta levantarlo (lo de Compose de
 *   siempre, `tryAwaitRelease`).
 */
object PressStep {
    /** Lo que le toca al dedo tras el movimiento. */
    sealed interface Move {
        /** Nada cambia: sigue con lo que tuviera (o sin nada). */
        data object Keep : Move

        /** Pasa a [bit]; `null` = se queda libre, sin nada pulsado. */
        data class To(val bit: Int?) : Move
    }

    /**
     * [held]: el bit que lleva ese dedo (null = ninguno). [over]: el bit que
     * hay bajo él ahora (null = ninguno).
     */
    fun next(slide: Boolean, sticky: Boolean, held: Int?, over: Int?): Move = when {
        slide -> if (over == held) Move.Keep else Move.To(over)
        sticky -> Move.Keep
        held != null && over != held -> Move.To(null)
        else -> Move.Keep
    }
}

File: ios/PepoMote/Control/PressStep.swift
import Foundation

/// Qué hacer con el dedo cuando se mueve, según los dos ajustes de pulsación.
/// Función pura calcada en las tres apps (Android `PressStep.kt`, Linux
/// `touch.rs`): la interfaz solo obedece lo que decide aquí.
enum PressStep: Equatable {
    /// El dedo sigue igual: ni suelta ni pulsa nada.
    case keep
    /// Soltar lo que tuviera y pulsar ese bit; `nil` deja el dedo libre.
    case to(UInt32?)

    /// `held`: lo que este dedo tiene pulsado ahora (`nil` = nada).
    /// `over`: el botón que le queda debajo (`nil` = ninguno).
    ///
    /// - «Pulsar deslizando»: el dedo va cogiendo lo que pisa y soltando lo
    ///   anterior (un dedo libre también pulsa al entrar en un botón).
    /// - Sin deslizar y con «Mantener al salir del botón»: nada cambia hasta
    ///   levantar el dedo.
    /// - Sin deslizar y sin mantener: salirse del botón lo suelta, y ese dedo
    ///   ya no vuelve a coger nada (como el `tryAwaitRelease` de Compose).
    static func next(slide: Bool, sticky: Bool, held: UInt32?, over: UInt32?) -> PressStep {
        if slide { return over == held ? .keep : .to(over) }
        if sticky { return .keep }
        if let held, over != held { return .to(nil) }
        return .keep
    }
}
